src/ripple/overlay/impl/Handshake.cpp

```diff
@@ -0,0 +1,302 @@
+    if (!cookie1)
+    {
+        JLOG (journal.error()) << "Cookie generation: local setup not complete";
+        return boost::none;
+    }
+
+    auto const cookie2 = hashLastMessage(
+        ssl.stream.native_handle(), SSL_get_peer_finished);
+    if (!cookie2)
+    {
+        JLOG (journal.error()) << "Cookie generation: peer setup not complete";
+        return boost::none;
+    }
+
+    auto const result = (*cookie1 ^ *cookie2);
+
+    // Both messages hash to the same value and the cookie
+    // is 0. Don't allow this.
+    if (result == beast::zero)
+    {
+        JLOG(journal.error()) << "Cookie generation: identical finished messages";
+        return boost::none;
+    }
+
+    return sha512Half (Slice (result.data(), result.size()));
+}
+
+void
+buildHandshake(
+    boost::beast::http::fields& h,
+    const ripple::uint256& sharedValue,
+    boost::optional<std::uint32_t> networkID,
+    beast::IP::Address public_ip,
+    beast::IP::Address remote_ip,
+    Application& app)
+{
+    if (networkID)
+    {
+        // The network identifier, if configured, can be used to specify
+        // what network we intend to connect to and detect if the remote
+        // end connects to the same network.
+        h.insert("Network-ID", std::to_string(*networkID));
+    }
+
+    h.insert ("Network-Time",
+        std::to_string(app.timeKeeper().now().time_since_epoch().count()));
+
+    h.insert ("Public-Key",
+        toBase58(TokenType::NodePublic, app.nodeIdentity().first));
+
+    {
+        auto const sig = signDigest(app.nodeIdentity().first,
+            app.nodeIdentity().second, sharedValue);
+        h.insert("Session-Signature",
+            base64_encode(sig.data(), sig.size()));
+    }
+
+    if (beast::IP::is_public (remote_ip))
+        h.insert ("Remote-IP", remote_ip.to_string());
+
+    if (!public_ip.is_unspecified())
+        h.insert ("Local-IP", public_ip.to_string());
+
+    if (auto const cl = app.getLedgerMaster().getClosedLedger())
+    {
+        // TODO: Use hex for these
+        h.insert ("Closed-Ledger", base64_encode(
+            cl->info().hash.begin(), cl->info().hash.size()));
+        h.insert ("Previous-Ledger", base64_encode(
+            cl->info().parentHash.begin(), cl->info().parentHash.size()));
+    }
+}
+
+PublicKey
+verifyHandshake(
+    boost::beast::http::fields const& headers,
+    const ripple::uint256& sharedValue,
+    boost::optional<std::uint32_t> networkID,
+    beast::IP::Address public_ip,
+    beast::IP::Address remote,
+    Application& app)
+{
+    if (networkID)
+    {
+        auto const iter = headers.find("Network-ID");
+
+        if (iter != headers.end())
+        {
+            std::uint32_t nid;
+
+            if (!beast::lexicalCastChecked(nid, iter->value().to_string()))
+                throw std::runtime_error("Invalid peer network identifier");
+
+            if(nid != *networkID)
+                throw std::runtime_error("Peer is on a different network");
```

Comment:
In general, we don't throw on "external input". This throws based on the peer protocol. I understand this is a different level of trust than an RPC command. Are we OK with this throw (and others in this function)?